File: Patterns/Manuscript/manuscript/code/java/DesignPatterns/src/main/java/facade/ControlPanel.java
package facade;

class ControlPanel {
    private AutomaticParkingSystem parkingSystem;
    private AdaptiveCruiseControlSystem cruiseControlSystem;
    private CollisionAvoidanceSystem collisionSystem;

    public ControlPanel() {
        this.parkingSystem = new AutomaticParkingSystem();
        this.cruiseControlSystem = new AdaptiveCruiseControlSystem();
        this.collisionSystem = new CollisionAvoidanceSystem();
    }

    public void park() {
        parkingSystem.park();
    }

    public void activateCruiseControl() {
        cruiseControlSystem.activate();
    }

    public void avoidCollision() {
        collisionSystem.avoid();
    }
}

File: code/csharp/DesignPatterns/facade/ControlPanel.cs
public class ControlPanel
{
    private IAutomaticParkingSystem _parkingSystem = new AutomaticParkingSystem();
    private IAdaptiveCruiseControlSystem _cruiseControlSystem = new AdaptiveCruiseControlSystem();
    private ICollisionAvoidanceSystem _collisionSystem = new CollisionAvoidanceSystem();

    public void Park()
    {
        _parkingSystem.Park();
    }

    public void ActivateCruiseControl()
    {
        _cruiseControlSystem.Activate();
    }

    public void AvoidCollision()
    {
        _collisionSystem.Avoid();
    }

    public void DriveToHomeAndPark()
    {
        System.Console.WriteLine("Taking you home...");
        ActivateCruiseControl();
        AvoidCollision();
        Park();
    }
}
